src/main/java/com/artipie/asto/s3/S3Storage.java

```diff
@@ -286,34 +289,59 @@ public S3Storage(final S3AsyncClient client, final String bucket, final boolean
         return content.size()
             .map(ignored -> CompletableFuture.completedFuture(content))
             .orElseGet(
-                () -> {
-                    final Flowable<ByteBuffer> cache = Flowable.fromPublisher(content).cache();
-                    return cache
-                        .map(Buffer::remaining)
-                        .scanWith(() -> 0L, (sum, item) -> sum + item)
-                        .takeUntil(total -> total >= limit)
-                        .lastOrError()
-                        .to(SingleInterop.get())
-                        .toCompletableFuture()
-                        .<Optional<Long>>thenApply(
-                            last -> {
-                                final Optional<Long> size;
-                                if (last >= limit) {
-                                    size = Optional.empty();
-                                } else {
-                                    size = Optional.of(last);
-                                }
-                                return size;
-                            }
-                        ).thenApply(
-                            sizeOpt -> sizeOpt
-                                .<Content>map(size -> new Content.From(size, cache))
-                                .orElse(new Content.From(cache))
+                () -> cacheInTmpFile(content).thenCompose(
+                    tmp -> {
+                        final Flowable<ByteBuffer> cache = Flowable.fromPublisher(
+                            new File(tmp).content()
                         );
-                }
+                        return cache
+                            .map(Buffer::remaining)
+                            .scanWith(() -> 0L, (sum, item) -> sum + item)
+                            .takeUntil(total -> total >= limit)
+                            .lastOrError()
+                            .to(SingleInterop.get())
+                            .toCompletableFuture()
+                            .<Optional<Long>>thenApply(
+                                last -> {
+                                    final Optional<Long> size;
+                                    if (last >= limit) {
+                                        size = Optional.empty();
+                                    } else {
+                                        size = Optional.of(last);
+                                    }
+                                    return size;
+                                }
+                            ).thenApply(
+                                sizeOpt -> {
+                                    final Flowable<ByteBuffer> data = cache.doAfterTerminate(
+                                        () -> Files.delete(tmp)
```

Comment:
@olegmoz should not the file be removed on the previous level terminate? Seem like it will not be removed if error occurs on scan for example

Reply:
@olenagerasimova the temporary file could be deleted only after all content has been read. But you are absolutely right that it should be cleaned up if something fails on when calculating size. Fixed it now, thank you